File: src/main/java/cc/mirukuneko/realtrainmodrenewed/client/screen/SignalValueScreen.java
package cc.mirukuneko.realtrainmodrenewed.client.screen;

import cc.mirukuneko.realtrainmodrenewed.blockentity.SignalRemoteBlockEntity;
import cc.mirukuneko.realtrainmodrenewed.client.ClientNetworkHelper;
import cc.mirukuneko.realtrainmodrenewed.network.BindSignalReceiverPayload;
import cc.mirukuneko.realtrainmodrenewed.network.SetSignalValuePayload;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;

public class SignalValueScreen extends Screen {
    private final BlockPos pos;
    private EditBox channelBox;
    private EditBox valueBox;
    private int linkedChannel = -1;

    public SignalValueScreen(BlockPos pos) {
        super(Component.literal("受信機(signal値)"));
        this.pos = pos.immutable();
    }

    @Override
    protected void init() {
        linkedChannel = readLinkedChannel();
        int boxWidth = 150;
        int x = (width - boxWidth) / 2;
        int baseY = height / 2 - 30;

        if (linkedChannel > 0) {
            valueBox = new EditBox(font, x, baseY, boxWidth, 20, Component.literal("signal値"));
            valueBox.setMaxLength(10);
            addRenderableWidget(valueBox);
            setInitialFocus(valueBox);
        } else {
            channelBox = new EditBox(font, x, baseY, boxWidth, 20, Component.literal("信号番号"));
            channelBox.setMaxLength(10);
            addRenderableWidget(channelBox);
            setInitialFocus(channelBox);
        }

        addRenderableWidget(Button.builder(Component.literal("決定"), button -> submit())
            .bounds(width / 2 - 75, baseY + 40, 70, 20)
            .build());
        addRenderableWidget(Button.builder(Component.translatable("gui.cancel"), button -> onClose())
            .bounds(width / 2 + 5, baseY + 40, 70, 20)
            .build());
    }

    private void submit() {
        try {
            if (linkedChannel > 0) {
                int signalValue = Integer.parseInt(valueBox.getValue().trim());
                ClientNetworkHelper.sendToServer(new SetSignalValuePayload(pos, signalValue));
            } else {
                int channel = Integer.parseInt(channelBox.getValue().trim());
                ClientNetworkHelper.sendToServer(new BindSignalReceiverPayload(pos, channel));
            }
            onClose();
        } catch (NumberFormatException ignored) {
            if (Minecraft.getInstance().player != null) {
                Minecraft.getInstance().player.sendOverlayMessage(Component.literal(
                    linkedChannel > 0 ? "signal値を入力してください" : "信号番号を入力してください"
                ));
            }
        }
    }

    private int readLinkedChannel() {
        Minecraft minecraft = Minecraft.getInstance();
        if (minecraft.level == null) {
            return -1;
        }
        if (minecraft.level.getBlockEntity(pos) instanceof SignalRemoteBlockEntity blockEntity) {
            return blockEntity.getLinkedChannel();
        }
        return -1;
    }

    @Override
    public void extractRenderState(GuiGraphicsExtractor graphics, int mouseX, int mouseY, float partialTick) {
        super.extractRenderState(graphics, mouseX, mouseY, partialTick);
        graphics.centeredText(font, title, width / 2, height / 2 - 70, 0xFFFFFF);
        if (linkedChannel > 0) {
            graphics.centeredText(font, Component.literal("登録済み信号番号: " + linkedChannel), width / 2, height / 2 - 58, 0xAAAAAA);
            graphics.centeredText(font, Component.literal("今回は signal値 のみ入力"), width / 2, height / 2 - 46, 0xAAAAAA);
        } else {
            graphics.centeredText(font, Component.literal("先に信号番号を登録"), width / 2, height / 2 - 58, 0xAAAAAA);
            graphics.centeredText(font, Component.literal("次回開いたときに signal値 を入力"), width / 2, height / 2 - 46, 0xAAAAAA);
        }
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/screen/SignalValueScreen.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.client.screen

import cc.mirukuneko.realtrainmodrenewed.blockentity.SignalRemoteBlockEntity
import cc.mirukuneko.realtrainmodrenewed.client.ClientNetworkHelper
import cc.mirukuneko.realtrainmodrenewed.network.BindSignalReceiverPayload
import cc.mirukuneko.realtrainmodrenewed.network.SetSignalValuePayload
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.Button
import net.minecraft.client.gui.components.EditBox
import net.minecraft.client.gui.screens.Screen
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component

open class SignalValueScreen(pos: BlockPos) : Screen(Component.translatable("block.realtrainmodrenewed.signal_value_receiver")) {
    private val pos: BlockPos = pos.immutable()
    private lateinit var channelBox: EditBox
    private lateinit var valueBox: EditBox
    private var linkedChannel: Int = -1

    override fun init() {
        linkedChannel = readLinkedChannel()
        val boxWidth = 150
        val x = (width - boxWidth) / 2
        val baseY = height / 2 - 30

        if (linkedChannel > 0) {
            valueBox = EditBox(font, x, baseY, boxWidth, 20, Component.translatable("screen.realtrainmodrenewed.signal.value"))
            valueBox.setMaxLength(10)
            addRenderableWidget(valueBox)
            setInitialFocus(valueBox)
        } else {
            channelBox = EditBox(font, x, baseY, boxWidth, 20, Component.translatable("screen.realtrainmodrenewed.signal.channel"))
            channelBox.setMaxLength(10)
            addRenderableWidget(channelBox)
            setInitialFocus(channelBox)
        }

        addRenderableWidget(
            Button.builder(Component.translatable("gui.done")) { submit() }
                .bounds(width / 2 - 75, baseY + 40, 70, 20)
                .build()
        )
        addRenderableWidget(
            Button.builder(Component.translatable("gui.cancel")) { onClose() }
                .bounds(width / 2 + 5, baseY + 40, 70, 20)
                .build()
        )
    }

    private fun submit() {
        try {
            if (linkedChannel > 0) {
                val signalValue = valueBox.value.trim().toInt()
                ClientNetworkHelper.sendToServer(SetSignalValuePayload(pos, signalValue))
            } else {
                val channel = channelBox.value.trim().toInt()
                ClientNetworkHelper.sendToServer(BindSignalReceiverPayload(pos, channel))
            }
            onClose()
        } catch (ignored: NumberFormatException) {
            if (Minecraft.getInstance().player != null) {
                Minecraft.getInstance().player!!.sendOverlayMessage(
                    Component.translatable(
                        if (linkedChannel > 0) {
                            "message.realtrainmodrenewed.signal.enter_value"
                        } else {
                            "message.realtrainmodrenewed.signal.enter_channel"
                        },
                    )
                )
            }
        }
    }

    private fun readLinkedChannel(): Int {
        val minecraft = Minecraft.getInstance()
        if (minecraft.level == null) {
            return -1
        }
        val blockEntity = minecraft.level!!.getBlockEntity(pos)
        if (blockEntity is SignalRemoteBlockEntity) {
            return blockEntity.linkedChannel
        }
        return -1
    }

    override fun extractRenderState(
        graphics: GuiGraphicsExtractor,
        mouseX: Int,
        mouseY: Int,
        partialTick: Float
    ) {
        super.extractRenderState(graphics, mouseX, mouseY, partialTick)
        graphics.centeredText(font, title, width / 2, height / 2 - 70, 0xFFFFFF)
        if (linkedChannel > 0) {
            graphics.centeredText(
                font,
                Component.translatable("screen.realtrainmodrenewed.signal_value.registered_channel", linkedChannel),
                width / 2,
                height / 2 - 58,
                0xAAAAAA
            )
            graphics.centeredText(
                font,
                Component.translatable("screen.realtrainmodrenewed.signal_value.enter_value_now"),
                width / 2,
                height / 2 - 46,
                0xAAAAAA
            )
        } else {
            graphics.centeredText(
                font,
                Component.translatable("screen.realtrainmodrenewed.signal_value.register_channel_first"),
                width / 2,
                height / 2 - 58,
                0xAAAAAA
            )
            graphics.centeredText(
                font,
                Component.translatable("screen.realtrainmodrenewed.signal_value.enter_value_next_time"),
                width / 2,
                height / 2 - 46,
                0xAAAAAA
            )
        }
    }

    override fun isPauseScreen(): Boolean = false
}
